### PR TITLE
DiscIO: Add GetRegion function and Region enum

### PR DESCRIPTION
**Pull request by JosJuice**
Instead of needing different switch cases for converting countries to regions in multiple places, we now only need a single country-to-region switch case (in DiscIO/Enums.cpp), and we get a nice Region type.

Additionally, the second commit replaces the bNTSC variable in SConfig with a variable of the new Region type, which lets us get rid of VideoInterface::SetRegionReg, a huge hack in CEXIMemoryCard, and some minor things.

### CODE REVIEW

File: Source/Core/Core/Boot/Boot.cpp [hunk 1]
```diff
@@ -255,7 +255,8 @@ bool CBoot::BootUp()
   g_symbolDB.Clear();
 
   // PAL Wii uses NTSC framerate and linecount in 60Hz modes
-  VideoInterface::Preset(_StartupPara.bNTSC || (_StartupPara.bWii && _StartupPara.bPAL60));
+  bool NTSC = DiscIO::IsNTSC(_StartupPara.m_region);
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot.cpp]: This should probably be left in the conditional, or both be moved to a new function. This variable is only ever used in one place (it's also non-const).

File: Source/Core/Core/Boot/Boot_BS2Emu.cpp [hunk 2]
```diff
@@ -74,8 +74,8 @@ bool CBoot::EmulatedBS2_GC(bool skipAppLoader)
       0x10000006,
       0x8000002C);  // Console type - DevKit  (retail ID == 0x00000003) see YAGCD 4.2.1.1.2
 
-  PowerPC::HostWrite_U32(SConfig::GetInstance().bNTSC ? 0 : 1,
-                         0x800000CC);  // Fake the VI Init of the IPL (YAGCD 4.2.1.4)
+  bool NTSC = DiscIO::IsNTSC(SConfig::GetInstance().m_region);
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot_BS2Emu.cpp]: `ntsc`. This should also be const.

File: Source/Core/Core/Boot/Boot_BS2Emu.cpp [hunk 3]
```diff
@@ -336,10 +324,7 @@ bool CBoot::EmulatedBS2_Wii()
   INFO_LOG(BOOT, "Faking Wii BS2...");
 
   // Setup Wii memory
-  DiscIO::Country country_code = DiscIO::Country::COUNTRY_UNKNOWN;
-  if (DVDInterface::VolumeIsValid())
-    country_code = DVDInterface::GetVolume().GetCountry();
-  if (SetupWiiMemory(country_code) == false)
+  if (SetupWiiMemory() == false)
```
Comment (lioncash) [inline Source/Core/Core/Boot/Boot_BS2Emu.cpp]: `!SetupWiiMemory()`

File: Source/Core/Core/BootManager.cpp [hunk 4]
```diff
@@ -349,17 +353,19 @@ bool BootCore(const std::string& _rFilename)
     g_SRAM_netplay_initialized = false;
   }
 
+  bool NTSC = DiscIO::IsNTSC(StartUp.m_region);
```
Comment (lioncash) [inline Source/Core/Core/BootManager.cpp]: `ntsc`, this should also be const.

File: Source/Core/Core/Boot/Boot_BS2Emu.cpp (hunk 3 above)
Comment (BhaaLseN) [inline Source/Core/Core/Boot/Boot_BS2Emu.cpp]: You could change that to `!SetupWiiMemory()` while you're at it ;)

File: Source/Core/Core/HW/EXI_DeviceMemoryCard.cpp [hunk 5]
```diff
@@ -152,73 +152,18 @@ CEXIMemoryCard::CEXIMemoryCard(const int index, bool gciFolder) : card_index(ind
...
-      break;
-    }
-  }
-  default:
-    country_code = DiscIO::Country::COUNTRY_EUROPE;
-    strDirectoryName += EUR_DIR DIR_SEP;
-  }
-  strDirectoryName += StringFromFormat("Card %c", 'A' + card_index);
+
+  bool shiftjis = region == DiscIO::Region::NTSC_J;
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI_DeviceMemoryCard.cpp]: `shift_jis`. this should also be const.

File: Source/Core/Core/HW/EXI_DeviceMemoryCard.cpp [hunk 6]
```diff
@@ -234,16 +179,15 @@ void CEXIMemoryCard::SetupGciFolder(u16 sizeMb)
     else  // we tried but the user wants to crash
     {
       // TODO more user friendly abort
-      PanicAlertT("%s is not a directory, failed to move to *.original.\n Verify your write "
-                  "permissions or move "
-                  "the file outside of Dolphin",
+      PanicAlertT("%s is not a directory, failed to move to *.original.\n Verify your "
+                  "write permissions or move the file outside of Dolphin",
                   strDirectoryName.c_str());
       exit(0);
```
Comment (lioncash) [inline Source/Core/Core/HW/EXI_DeviceMemoryCard.cpp]: Not a review comment, but just lol @ this.

File: Source/Core/Core/HW/GCMemcard.h [hunk 7]
```diff
@@ -126,11 +126,11 @@ struct Header  // Offset    Size    Description
   // Nintendo format algorithm.
   // Constants are fixed by the GC SDK
   // Changing the constants will break memory card support
-  Header(int slot = 0, u16 sizeMb = MemCard2043Mb, bool ascii = true)
+  Header(int slot = 0, u16 sizeMb = MemCard2043Mb, bool shiftjis = false)
```
Comment (lioncash) [inline Source/Core/Core/HW/GCMemcard.h]: `shift_jis`

File: Source/Core/Core/Boot/Boot_BS2Emu.cpp (hunk 3 above)
Comment (JosJuice) [inline Source/Core/Core/Boot/Boot_BS2Emu.cpp]: Done.

File: Source/Core/DiscIO/Enums.h [hunk 8]
```diff
@@ -38,6 +38,16 @@ enum class Country
   NUMBER_OF_COUNTRIES
 };
 
+// Regions 0 - 2 and 4 match Nintendo's Wii region numbering.
+enum class Region
+{
+  NTSC_J = 0,      // Japan and Taiwan (and for GC only, South Korea?)
+  NTSC_U = 1,      // Mainly North America
+  PAL = 2,         // Mainly Europe and Oceania
+  UNKNOWN_REGION,  // 3 seems to be unused? Anyway, we need an UNKNOWN_REGION. Let's put it here
```
Comment (lioncash) [inline Source/Core/DiscIO/Enums.h]: Should probably put an `= 3,` here to make it consistent.

File: Source/Core/DiscIO/VolumeDirectory.cpp [hunk 9]
```diff
@@ -164,6 +164,14 @@ void CVolumeDirectory::SetGameID(const std::string& id)
   memcpy(m_diskHeader.data(), id.c_str(), std::min(id.length(), MAX_ID_LENGTH));
 }
 
+Region CVolumeDirectory::GetRegion() const
+{
+  if (m_is_wii)
+    return RegionSwitchWii(m_diskHeader[3]);
+  else
```
Comment (lioncash) [inline Source/Core/DiscIO/VolumeDirectory.cpp]: `else` isn't necessary here.

**Comment by JosJuice** [comment]
PR #4592 brought some new information to my intention. I've updated the comment in GCMemcardDirectory.cpp, but the relevant code is the same as in earlier versions of this PR.

**Comment by JMC47** [comment]
This is causing the Homebrew Channel to crash on launch.

**Pull request merged**